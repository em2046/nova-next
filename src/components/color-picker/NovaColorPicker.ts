import {
  computed,
  defineComponent,
  h,
  reactive,
  VNode,
  watch,
  onMounted,
} from 'vue';
import Color from './color';
import Utils from '../../utils/utils';
import { MousePosition } from '../../uses/useMousemove';
import HsvPanel from './parts/HsvPanel';
import HueSlide from './parts/slides/HueSlide';
import AlphaSlide from './parts/slides/AlphaSlide';
import Preview from './parts/Preview';
import RgbaLabels from './parts/labels/RgbaLabels';
import HexLabel from './parts/labels/HexLabel';

export default defineComponent({
  model: {
    event: 'update',
  },
  props: {
    value: {
      type: String,
      default: '#ff0000',
    },
  },
  setup: function (props, context) {
    const emit = context.emit;

    const state = reactive({
      position: {
        // pos [0, 200] -> CSS h [0, 360]
        hue: 0,
        // pos [0, 200] -> CSS s [0, 100]
        saturation: 200,
        // pos [0, 200] -> CSS v [100, 0]
        value: 0,
        // pos [0, 200] -> CSS a [1, 0]
        alpha: 0,
      },
      color: Color.fromCssLikeHsva(0, 100, 100, 1),
    });

    const hueDegrees = computed(() => {
      return Math.round((state.position.hue / 200) * 360) % 360;
    });

    function getColorFromPosition(): Color {
      return Color.fromCssLikeHsva(
        hueDegrees.value,
        state.position.saturation / 2,
        (200 - state.position.value) / 2,
        (200 - state.position.alpha) / 200
      );
    }

    function setPositionFromColor(color: Color): void {
      const { h, s, v, a } = color.toHsva();

      const hue = Utils.numberLimit((h / 360) * 200, 0, 200);
      const saturation = Utils.numberLimit(s * 200, 0, 200);
      const value = Utils.numberLimit(200 - 200 * v, 0, 200);
      const alpha = Utils.numberLimit(200 - 200 * a, 0, 200);

      state.position.hue = hue;
      state.position.saturation = saturation;
      state.position.value = value;
      state.position.alpha = alpha;
    }

    function setColor(color: Color): void {
      state.color = color;
    }

    function setColorAndPosition(color: Color): void {
      const panelColor = getColorFromPosition();

      if (panelColor.toHex() !== color.toHex()) {
        setPositionFromColor(color);
        setColor(color);
      }
    }

    function setColorFromPosition(): void {
      setColor(getColorFromPosition());
    }

    function updatePropsValue(color: Color): void {
      emit('update', `#${color.toHex()}`);
    }

    watch(
      () => props.value,
      (value, prevValue) => {
        if (value !== prevValue) {
          const color = Color.fromHex(value);
          setColorAndPosition(color);
        }
      }
    );

    onMounted(() => {
      const color = Color.fromHex(props.value);
      setColorAndPosition(color);
    });

    return (): VNode => {
      const hsvPanelNode = h(HsvPanel, {
        hueReg: hueDegrees.value,
        saturation: state.position.saturation,
        value: state.position.value,
        onMove: (position: MousePosition) => {
          state.position.saturation = Utils.numberLimit(position.x, 0, 200);
          state.position.value = Utils.numberLimit(position.y, 0, 200);
          setColorFromPosition();
        },
        onFinish: () => {
          updatePropsValue(state.color);
        },
      });

      const hueSlideNode = h(HueSlide, {
        hue: state.position.hue,
        onMove: (position: MousePosition) => {
          state.position.hue = Utils.numberLimit(position.y, 0, 200);
          setColorFromPosition();
        },
        onFinish: () => {
          updatePropsValue(state.color);
        },
      });

      const alphaSlideNode = h(AlphaSlide, {
        alpha: state.position.alpha,
        color: state.color,
        onMove: (position: MousePosition) => {
          state.position.alpha = Utils.numberLimit(position.y, 0, 200);
          setColorFromPosition();
        },
        onFinish: () => {
          updatePropsValue(state.color);
        },
      });

      const slidesNode = h('div', { class: 'nova-color-picker-slides' }, [
        hueSlideNode,
        alphaSlideNode,
      ]);

      const formNode = h('div', { class: 'nova-color-picker-form' }, [
        h(RgbaLabels, {
          color: state.color,
          onColorInput: (color: Color) => {
            setColorAndPosition(color);
          },
          onColorBlur: (color: Color) => {
            updatePropsValue(color);
          },
        }),
        h(HexLabel, {
          color: state.color,
          onColorInput: (color: Color) => {
            setColorAndPosition(color);
          },
          onColorBlur: (color: Color) => {
            updatePropsValue(color);
          },
        }),
      ]);

      const previewNode = h(Preview, {
        color: state.color,
      });

      return h(
        'div',
        { class: 'nova-color-picker' },
        h('div', { class: 'nova-color-picker-panel' }, [
          hsvPanelNode,
          slidesNode,
          formNode,
          previewNode,
        ])
      );
    };
  },
});
