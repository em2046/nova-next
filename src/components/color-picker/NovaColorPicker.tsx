import {
  computed,
  onMounted,
  reactive,
  ref,
  SetupContext,
  VNodeProps,
  watch,
} from 'vue';
import { vueJsxCompat } from '../../vue-jsx-compat';
import { MovePosition } from '../../uses/use-move';
import Utils from '../../utils/utils';
import Color, { ColorFormat } from './color';
import { Trigger } from './parts/Trigger';
import { HsvPanel } from './parts/HsvPanel';
import { HueSlide } from './parts/slides/HueSlide';
import { AlphaSlide } from './parts/slides/AlphaSlide';
import { RgbaLabels } from './parts/labels/RgbaLabels';
import { HslaLabels } from './parts/labels/HslaLabels';
import { HexLabel } from './parts/labels/HexLabel';
import { Preview } from './parts/Preview';
import { PresetValues } from './parts/PresetValues';
import useEnvironment, {
  environmentProps,
  NovaEnvironmentProps,
} from '../../uses/use-environment';
import { MDIClose } from '@em2046/material-design-icons-vue-next';
import { NovaDropdown } from '../dropdown';
import {
  DropdownInstance,
  dropdownProps,
  NovaDropdownProps,
  NovaDropdownTriggerScoped,
} from '../dropdown/NovaDropdown';

//region Mode
const modeRgba = Symbol('rgba');
const modeHsla = Symbol('hsla');

const labelsMap = new Map([
  [modeRgba, RgbaLabels],
  [modeHsla, HslaLabels],
]);

const modeList = [...labelsMap.keys()];
const modeSize = modeList.length;

//endregion

export interface NovaColorPickerProps
  extends NovaEnvironmentProps,
    NovaDropdownProps {
  value?: string;
  alpha?: boolean;
  format?: ColorFormat;
  preset?: string[];
  onUpdate?: (color: string) => void;
  onOpenChange?: (opened: boolean) => void;
}

const defaultValue = '#ff0000';

const colorPickerProps = {
  ...environmentProps,
  ...dropdownProps,
  value: {
    type: String,
    default: defaultValue,
  },
  alpha: {
    type: Boolean,
    default: false,
  },
  format: {
    type: String,
    default: 'hex',
  },
  preset: {
    type: Array,
    default: null,
  },
};

export interface NovaColorPickerPresetScoped {
  preset: string[];
  color: Color;
  setColorAndPosition: (color: Color) => void;
}

export interface NovaColorPickerTriggerScoped
  extends NovaDropdownTriggerScoped {
  color: Color;
}

const NovaColorPickerImpl = {
  name: 'NovaColorPicker',
  props: colorPickerProps,
  setup(props: NovaColorPickerProps, context: SetupContext) {
    const emit = context.emit;

    const environment = useEnvironment(props as NovaEnvironmentProps);
    const dropdownInstanceRef = ref<DropdownInstance | null>(null);

    const mode = props.format === 'hsl' ? modeHsla : modeRgba;
    const state = reactive({
      position: {
        /**
         * [0, 200] -> CSS [0, 360]
         */
        hue: 0,
        /**
         * [0, 200] -> CSS [0, 100]
         */
        saturation: 200,
        /**
         * [0, 200] -> CSS [100, 0]
         */
        value: 0,
        /**
         * [0, 200] -> CSS [1, 0]
         */
        alpha: 0,
      },
      color: Color.fromCssLikeHsva(0, 100, 100, 1),
      mode,
    });

    /**
     * [0, 360]
     */
    const hueDegrees = computed(() => {
      return Math.round((state.position.hue / 200) * 360) % 360;
    });

    const classList = computed(() => {
      return [
        'nova-color-picker',
        {
          ['nova-color-picker-disabled']: props.disabled,
        },
      ];
    });

    const dropdownClassList = computed(() => {
      return [
        props.dropdownClass,
        {
          ['nova-color-picker-panel-has-alpha']: props.alpha,
        },
      ];
    });

    function getColorFromPosition(): Color {
      return Color.fromCssLikeHsva(
        hueDegrees.value,
        state.position.saturation / 2,
        (200 - state.position.value) / 2,
        (200 - state.position.alpha) / 200
      );
    }

    function setPositionFromColor(color: Color): void {
      const hsva = color.toHsva();

      const hue = Utils.numberLimit((hsva.hue / 360) * 200, 0, 200);
      const saturation = Utils.numberLimit(hsva.saturation * 200, 0, 200);
      const value = Utils.numberLimit(200 - 200 * hsva.value, 0, 200);
      const alpha = Utils.numberLimit(200 - 200 * hsva.alpha, 0, 200);

      state.position.hue = hue;
      state.position.saturation = saturation;
      state.position.value = value;
      state.position.alpha = alpha;
    }

    function setColor(color: Color): void {
      if (props.alpha) {
        state.color = color;
      } else {
        state.color = new Color(color.red, color.green, color.blue);
      }
    }

    function setColorAndPosition(color: Color): void {
      setPositionFromColor(color);
      setColor(color);
    }

    function setColorFromPosition(): void {
      setColor(getColorFromPosition());
    }

    function changePropsValue(color: Color): void {
      // JSX onUpdate
      emit('update', color.toString(props.format));

      // Template v-model:value
      emit('update:value', color.toString(props.format));
    }

    function switchMode(): void {
      let activeModeIndex = modeList.findIndex((mode) => {
        return state.mode === mode;
      });

      activeModeIndex++;
      if (activeModeIndex >= modeSize) {
        activeModeIndex = 0;
      }

      state.mode = modeList[activeModeIndex];
    }

    function onOpenChange(open: boolean) {
      if (!open) {
        changePropsValue(state.color);
      }
    }

    watch(
      () => props.value,
      (value) => {
        const color = Color.parse(value);
        setColorAndPosition(color);
      }
    );

    watch(
      () => props.alpha,
      (value) => {
        if (!value) {
          const { red, green, blue } = state.color;
          const color = new Color(red, green, blue);
          setColorAndPosition(color);
        }
      }
    );

    function reset() {
      const color = Color.parse(props.value);
      setColorAndPosition(color);
    }

    function init(): void {
      reset();
    }

    onMounted(() => {
      init();
    });

    return (): JSX.Element => {
      const language = environment.languageRef.value.colorPicker;

      function createTrigger() {
        const triggerProps = {
          disabled: !!props.disabled,
          color: state.color,
          environment,
        };

        const trigger = context.slots.trigger;

        let triggerNode = null;
        if (trigger) {
          triggerNode = () =>
            trigger({
              ...triggerProps,
            });
        }

        return <Trigger {...triggerProps}>{triggerNode}</Trigger>;
      }

      function createHsvPanel() {
        function onHsvMove(position: MovePosition): void {
          state.position.saturation = Utils.numberLimit(position.x, 0, 200);
          state.position.value = Utils.numberLimit(position.y, 0, 200);
          setColorFromPosition();
        }

        return (
          <HsvPanel
            hueReg={hueDegrees.value}
            saturation={state.position.saturation}
            value={state.position.value}
            onMove={onHsvMove}
          />
        );
      }

      function createHue() {
        function onHueMove(position: MovePosition): void {
          state.position.hue = Utils.numberLimit(position.y, 0, 200);
          setColorFromPosition();
        }

        return <HueSlide hue={state.position.hue} onMove={onHueMove} />;
      }

      function createAlpha() {
        if (!props.alpha) {
          return null;
        }

        function onAlphaMove(position: MovePosition): void {
          state.position.alpha = Utils.numberLimit(position.y, 0, 200);
          setColorFromPosition();
        }

        return (
          <AlphaSlide
            alpha={state.position.alpha}
            color={state.color}
            onMove={onAlphaMove}
          />
        );
      }

      function createSlides() {
        const alphaSlideNode = createAlpha();
        const hueSlideNode = createHue();

        return (
          <div class="nova-color-picker-slides">
            {hueSlideNode}
            {alphaSlideNode}
          </div>
        );
      }

      function createLabels() {
        const CurrLabels = labelsMap.get(state.mode);
        if (!CurrLabels) {
          return null;
        }

        return (
          <CurrLabels
            color={state.color}
            alpha={!!props.alpha}
            onColorInput={setColorAndPosition}
            onColorBlur={setColorAndPosition}
            environment={environment}
          />
        );
      }

      function createForm() {
        function onKeydown(e: KeyboardEvent) {
          switch (e.key) {
            case 'Enter':
              switchMode();
              e.stopPropagation();
              break;
          }
        }

        const labelsNode = createLabels();
        return (
          <div class="nova-color-picker-form">
            <div
              class="nova-color-picker-labels-switch"
              role="button"
              aria-label={language.aria.switch}
              onClick={switchMode}
              onKeydown={onKeydown}
              tabindex={0}
            />
            {labelsNode}
            <HexLabel
              color={state.color}
              onColorInput={setColorAndPosition}
              onColorBlur={setColorAndPosition}
            />
          </div>
        );
      }

      function createPreview() {
        return (
          <Preview
            color={state.color}
            value={props.value || defaultValue}
            onReset={init}
          />
        );
      }

      function createPreset() {
        const slotPreset = context.slots.preset;
        const presetProps = {
          preset: props.preset,
          color: state.color,
        };

        let slotPresetNode = null;
        if (slotPreset) {
          slotPresetNode = () =>
            slotPreset({
              ...presetProps,
              setColorAndPosition,
            });
        }

        if (!props.preset?.length && !slotPresetNode) {
          return null;
        }

        return (
          <PresetValues
            preset={props.preset || []}
            color={state.color}
            onSelect={setColorAndPosition}
          >
            {slotPresetNode}
          </PresetValues>
        );
      }

      function createDropdown() {
        const hsvPanelNode = createHsvPanel();
        const slidesNode = createSlides();
        const formNode = createForm();
        const previewNode = createPreview();
        const presetNode = createPreset();

        function closeDropdown() {
          dropdownInstanceRef.value?.close();
        }

        return (
          <div class="nova-color-picker-panel-inner">
            {hsvPanelNode}
            {slidesNode}
            {formNode}
            {previewNode}
            {presetNode}
            <div
              class="nova-color-picker-close"
              tabindex={0}
              role="button"
              aria-label={language.aria.close}
              onClick={closeDropdown}
            >
              <MDIClose />
            </div>
          </div>
        );
      }

      const triggerNode = createTrigger();
      const dropdownNode = createDropdown();

      const slots = {
        trigger: ({ dropdownInstance }: NovaDropdownTriggerScoped) => {
          dropdownInstanceRef.value = dropdownInstance;
          return triggerNode;
        },
        default: () => {
          return dropdownNode;
        },
      };

      return (
        <NovaDropdown
          class={classList.value}
          disabled={props.disabled}
          dropdownClass={dropdownClassList.value}
          dropdownStyle={props.dropdownStyle}
          dropdownProps={props.dropdownProps}
          teleportToBody={props.teleportToBody}
          placement={props.placement}
          environment={environment}
          onOpenChange={onOpenChange}
        >
          {slots}
        </NovaDropdown>
      );
    };
  },
};

export const NovaColorPicker = (NovaColorPickerImpl as unknown) as {
  new (): {
    $props: VNodeProps & NovaColorPickerProps;
  };
};
