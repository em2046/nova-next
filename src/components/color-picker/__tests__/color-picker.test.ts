import { mount } from '@vue/test-utils';
import NovaColorPicker from '../NovaColorPicker';
import { h, VNode, reactive } from 'vue';

describe('color-picker', () => {
  test('render', async () => {
    const wrapper = mount({
      setup() {
        const state = reactive({
          color: '#7f7f7f7f',
        });
        return (): VNode => {
          return h('div', [
            h(NovaColorPicker, {
              value: state.color,
              teleportToBody: false,
            }),
          ]);
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.find('.nova-color-picker-trigger').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
