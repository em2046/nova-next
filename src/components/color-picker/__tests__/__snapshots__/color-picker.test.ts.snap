// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`color-picker render 1`] = `"<div><div class=\\"nova-color-picker\\"><div class=\\"nova-color-picker-trigger\\"><div class=\\"nova-color-picker-trigger-inner\\" style=\\"background-color: rgb(255, 0, 0);\\"></div></div><!----></div></div>"`;

exports[`color-picker render 2`] = `"<div><div class=\\"nova-color-picker\\"><div class=\\"nova-color-picker-trigger\\"><div class=\\"nova-color-picker-trigger-inner\\" style=\\"background-color: rgba(128, 128, 128, 0.5);\\"></div></div><!--teleport start--><div class=\\"nova-color-picker-panel\\" style=\\"left: 0px; top: 0px;\\"><div class=\\"nova-color-picker-hsv\\" style=\\"background-color: rgb(255, 0, 0);\\"><div class=\\"nova-color-picker-saturation\\"></div><div class=\\"nova-color-picker-value\\"></div><div class=\\"nova-color-picker-cursor\\" style=\\"transform: translate(0px, 99.61px);\\"></div></div><div class=\\"nova-color-picker-slides\\"><div class=\\"nova-color-picker-hue-slide\\"><div class=\\"nova-color-picker-hue-bar\\"></div><div class=\\"nova-color-picker-hue-thumb\\" style=\\"transform: translate(0, 0px);\\"></div></div><div class=\\"nova-color-picker-alpha-slide\\"><div class=\\"nova-color-picker-alpha-bar\\"></div><div class=\\"nova-color-picker-alpha-thumb\\" style=\\"transform: translate(0, 99.61px);\\"></div></div></div><div class=\\"nova-color-picker-form\\"><div class=\\"nova-color-picker-labels\\"><label class=\\"nova-color-picker-label\\"><div class=\\"nova-color-picker-label-text\\">R</div><div class=\\"nova-color-picker-number\\"><input></div></label><label class=\\"nova-color-picker-label\\"><div class=\\"nova-color-picker-label-text\\">G</div><div class=\\"nova-color-picker-number\\"><input></div></label><label class=\\"nova-color-picker-label\\"><div class=\\"nova-color-picker-label-text\\">B</div><div class=\\"nova-color-picker-number\\"><input></div></label><label class=\\"nova-color-picker-label\\"><div class=\\"nova-color-picker-label-text\\">A</div><div class=\\"nova-color-picker-number\\"><input></div></label></div><div class=\\"nova-color-picker-output\\"><div class=\\"nova-color-picker-hex\\"><input></div></div><div class=\\"nova-color-picker-labels-switch\\"></div></div><div class=\\"nova-color-picker-preview\\"><div class=\\"nova-color-picker-preview-fill-left\\"></div><div class=\\"nova-color-picker-preview-fill-right\\"></div><div class=\\"nova-color-picker-preview-prev\\" style=\\"background-color: rgba(128, 128, 128, 0.5);\\"></div><div class=\\"nova-color-picker-preview-curr\\" style=\\"background-color: rgba(128, 128, 128, 0.5);\\"></div></div></div><!--teleport end--></div></div>"`;
